### PR TITLE
Add warnings to extensions

### PR DESCRIPTION
**Pull request by itaykl**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -95,6 +96,52 @@ const server = new PolarisServer({
...
 
+### Warnings
+
+In order to have the ability of warnings, which returned in the extensions of the response, you will need to supply them to
+polaris. you can supply the warnings through the `customContext` - when you create the `customContext`(of type `PolarisGraphQLContext`),
+set the `returnedExtensions` with the relevant warnings. let's see an example:
+
+```typescript
+const warnings: PolarisWarning[] = ['warning 1', 'warning 2'];
+const extensions: PolarisExtensions = { warnings: warnings };
```
Comment (osher-sade) [inline README.md]: you should add your warnings like this:
`customContext.returnedExtensions.warnings.push('warning 1')`
they should not set the whole returned extensions field, just the warnings

Comment (osher-sade) [reply README.md]: Show example of a resolver that adds a warning, which is not using custom context.
Use normal context, to make it less complicated.

File: src/config/polaris-server-config.ts [hunk 2]
```diff
@@ -23,4 +23,5 @@ export interface PolarisServerConfig extends ApolloServerExpressConfig {
     customContext?: (context: ExpressContext) => any;
     connection?: PolarisConnection;
     supportedRealities?: RealitiesHolder;
+    shouldAddWarningsToExtensions?: boolean;
```
Comment (osher-sade) [inline src/config/polaris-server-config.ts]: This should not be optional

File: src/server/polaris-server.ts [hunk 3]
```diff
@@ -98,8 +98,16 @@ export class PolarisServer {
     }
 
     private getApolloServerConfigurations(): ApolloServerExpressConfig {
+        const shouldAddWarningsToExtensions: boolean =
```
Comment (osher-sade) [inline src/server/polaris-server.ts]: This logic should be in `getActualConfig` method

File: test/integration-tests/test-server/test-server.ts [hunk 4]
```diff
@@ -27,15 +26,30 @@ const customContext = (context: ExpressContext): Partial<TestContext> => {
...
+    } else {
+        return customContext(context);
+    }
+};
+
+export async function startTestServer(
+    customCtx?: Partial<TestContext>,
+    shouldAddWarningsToExtensions?: boolean,
+): Promise<PolarisServer> {
     jest.setTimeout(150000);
```
Comment (osher-sade) [inline test/integration-tests/test-server/test-server.ts]: This should not be here!

File: test/integration-tests/test-server/test-server.ts [hunk 5]
```diff
@@ -27,15 +26,30 @@ const customContext = (context: ExpressContext): Partial<TestContext> => {
     };
 };
 
-export async function startTestServer(): Promise<PolarisServer> {
+const getCustomContext = (
```
Comment (osher-sade) [inline test/integration-tests/test-server/test-server.ts]: Why do you need this method?

File: test/integration-tests/test-server/test-server.ts [hunk 6]
```diff
@@ -27,15 +26,30 @@ const customContext = (context: ExpressContext): Partial<TestContext> => {
...
+    customCtx?: Partial<TestContext>,
+): Partial<TestContext> => {
+    if (customCtx) {
+        return customCtx;
+    } else {
+        return customContext(context);
+    }
+};
+
+export async function startTestServer(
```
Comment (osher-sade) [inline test/integration-tests/test-server/test-server.ts]: `startTestServer` should not accept these arguments

Comment (itaykl) [reply test/integration-tests/test-server/test-server.ts]: it's optional and i have to pass them here in order to have the ability to create a server with custom context that contains the warnings. also its the test server and i think it is legit

File: test/integration-tests/tests/warnings.test.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import { initializeDatabase } from '../test-server/data-initalizer';
+import { graphqlRawRequest } from '../test-server/graphql-client';
+import { startTestServer, stopTestServer } from '../test-server/test-server';
+import * as simpleQuery from './jsonRequestsAndHeaders/simpleQuery.json';
+
+let polarisServer: PolarisServer;
+
+describe('warnings tests - shouldAddWarningsToExtensions is true', () => {
+    beforeEach(async () => {
+        const customWarnings: PolarisWarning[] = ['warning 1', 'warning 2'];
```
Comment (osher-sade) [inline test/integration-tests/tests/warnings.test.ts]: This should not be in the before.
The way you should test this logic, is to activate a resolver that adds warning, and check that the extensions in the response contains warnings.

File: test/integration-tests/test-server/test-server.ts (hunk 5 above)
Comment (itaykl) [reply test/integration-tests/test-server/test-server.ts]: i need this method to distinguish between the tests with the custom context with warnings and the rest

File: src/server/polaris-server.ts [hunk 8]
```diff
@@ -98,8 +98,16 @@ export class PolarisServer {
...
+            this.polarisServerConfig.shouldAddWarningsToExtensions === undefined
+                ? false
```
Comment (osher-sade) [inline src/server/polaris-server.ts]: Default value should be `true`

File: README.md [hunk 9]
```diff
@@ -41,6 +41,7 @@ Through this interface you should set the following configurations which will be
     If you do not provide this property, the core will use default values for the logger.
 -   **middlewareConfiguration** (_MiddlewareConfiguration - optional_) - This is an interface that defines what core middlewares should be activated/disabled.
 -   **connection** (_Connection - optional_) - This class represents your connection with the database. Used in the core middlewares.
+-   **shouldAddWarningsToExtensions** (_boolean - optional_) - _Default: false._ Specifies whether to return the warnings in the response extensions or not.
```
Comment (osher-sade) [inline README.md]: Default value should be `true`

File: README.md [hunk 10]
```diff
@@ -175,6 +177,58 @@ const customContext = (context: ExpressContext): Partial<CustomContext> => {
 };
 ```
 
+### Warnings
+
+In order to have the ability of warnings, which returned in the extensions of the response, you will need to supply them to
+polaris. you can supply the warnings through the `customContext` - when you create the `customContext`(of type `PolarisGraphQLContext`),
```
Comment (osher-sade) [inline README.md]: through the context, and remove the "when" part

File: src/server/polaris-server.ts [hunk 11]
```diff
@@ -46,6 +43,10 @@ export class PolarisServer {
             logger: config.logger || getDefaultLoggerConfiguration(),
             applicationProperties: config.applicationProperties || { version: 'v1' },
             allowSubscription: config.allowSubscription || false,
+            shouldAddWarningsToExtensions:
```
Comment (osher-sade) [inline src/server/polaris-server.ts:46]: change to 
`shouldAddWarningsToExtensions: config.shouldAddWarningsToExtensions || true`

File: test/integration-tests/test-server/test-server.ts [hunk 12]
```diff
@@ -46,6 +46,27 @@ export async function startTestServer(): Promise<PolarisServer> {
     return server;
 }
 
+export async function startTestServerWithWarnings(
```
Comment (osher-sade) [inline test/integration-tests/test-server/test-server.ts]: rename to `startTestServerWithoutWarnings`

File: test/integration-tests/test-server/test-server.ts [hunk 13]
```diff
@@ -46,6 +46,27 @@ export async function startTestServer(): Promise<PolarisServer> {
...
+    config?: Partial<PolarisServerOptions>,
```
Comment (osher-sade) [inline test/integration-tests/test-server/test-server.ts]: Make it so it will recieve just a boolean

File: src/server/polaris-server.ts (hunk 11 above)
Comment (itaykl) [reply src/server/polaris-server.ts:46]: it's not good because when it's false it converts to true becuse of the 'or'

**Review by osher-sade — APPROVED** [review]

**Pull request merged**